src/storage/src/source/probe.rs

```diff
@@ -0,0 +1,96 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Support for sending frontier probes to upstream systems.
+
+use std::time::Duration;
+
+use mz_ore::now::{EpochMillis, NowFn};
+use mz_repr::Timestamp;
+use tracing::trace;
+
+/// A ticker to drive source upstream probing.
+///
+/// This type works similar to [`tokio::time::Interval`] but returns timestamps from its
+/// [`Ticker::tick`] method that can be used as probe timestamps. These timestamps are rounded down
+/// to the nearest multiple of the tick interval, to reduce the amount of unique timestamps emitted
+/// by sources, thereby reducing churn in downstream dataflows.
+///
+/// The ticker also supports usage in non-async contexts, using [`Ticker::tick_blocking`].
+pub(super) struct Ticker {
+    interval: EpochMillis,
+    now: NowFn,
+    last_tick: Option<EpochMillis>,
+}
+
+impl Ticker {
+    pub fn new(interval: Duration, now: NowFn) -> Self {
+        let interval = interval.as_millis().try_into().unwrap();
+        Self {
+            interval,
+            now,
+            last_tick: None,
+        }
+    }
+
+    /// Wait until it is time for the next probe, returning a suitable probe timestamp.
+    ///
+    /// This method tries to resolve as close as possible to the returned probe timestamp, though
+    /// it is not guaranteed to always succeed. If a tick is missed, it is skipped entirely.
+    pub async fn tick(&mut self) -> Timestamp {
+        let target = self.next_tick_target();
+
+        let mut now = (self.now)();
+        while now < target {
+            let wait = Duration::from_millis(target - now);
+            tokio::time::sleep(wait).await;
+            now = (self.now)();
+        }
+
+        trace!(target, now, "probe ticker skew: {}ms", now - target);
+        self.apply_tick(now)
+    }
+
+    /// Wait until it is time for the next probe, returning a suitable probe timestamp.
+    ///
+    /// Blocking version of [`Ticker::tick`].
+    pub fn tick_blocking(&mut self) -> Timestamp {
+        let target = self.next_tick_target();
+
+        let mut now = (self.now)();
+        while now < target {
+            let wait = Duration::from_millis(target - now);
+            std::thread::sleep(wait);
+            now = (self.now)();
+        }
+
+        trace!(target, now, "probe ticker skew: {}ms", now - target);
+        self.apply_tick(now)
+    }
+
+    /// Return the desired time of the next tick.
+    fn next_tick_target(&self) -> EpochMillis {
+        let target = match self.last_tick {
+            Some(ms) => ms + self.interval,
```

Comment:
Might this over time get out of sync with the `NowFn`/wall-clock time? If we only ever keep adding an interval to a timestamp we got initially.

Reply:
Ah, but we're not doing that! We call the `NowFn` in each tick and set the returned timestamp (rounded down) as `last_tick` to which the interval is then added.